### DISCUSSION TITLE
Dealing with electron energy density

### DISCUSSION THREAD

**Discussion by keniley1** [shannon-lab/zapdos#90] [General / Ideas]
Hey all,

Since we have gone full AD, I think it would be a relatively simple exercise to make the mean electron energy a completely independent variable. Right now it's represented with a variable called "mean_en" which is actually the log-molar energy density, n_e*2/3*Te. As far as I can tell, the primary motivation behind this was allowing the electron energy equation to use the same kernels as the density drift-diffusion terms. This is admittedly really nice, but I can think of a couple reasons to decouple them entirely. 

1. I need to have specific kernels and materials in Crane to deal with the fact that the mean electron energy is represented with std::exp(_em - _mean_en). It would obviously be much simpler if it was just a single generic variable, and the user could select which nonlinear variable they would like to use to interpolate rate coefficients. This would keep Crane completely abstract and free of any Zapdos-specific code.
2. There might be situations where the user would want to specify the electron temperature somewhere, like at an inlet when they don't want to model the actual discharge and only the afterglow physics. I've had a couple students email me trying to do this and run into trouble trying to wrap their heads around the log-mole energy density. 

The first point can admittedly be addressed much more simply: we could create a Material class that creates a property called "Te" and computes std::exp(_em - _mean_en). (The DriftDiffusionAction could add this automatically.) That way the user could select either a variable or a material property to interpolate rate coefficients, the jacobians would be added correctly, no additional kernels are needed, and I could keep Crane completely abstract. It would also allow AD to use ParsedFunctions correctly if the rate coefficients are in Arrhenius form. (Right now we usually create an AuxVariable called Te, which means the jacobians aren't being computed if Te is used in a function-based rate coefficient.) This doesn't address the second point at all, but I think good documentation and examples would solve that issue anyway. 

Any thoughts? 

-Shane

**Comment by csdechant** [comment]
I think added the Material class to create a property called "Te" might be best (unless users wanted new physics that requires only the electron energy and not the electron energy density). 

For the second point, are you talking about a DirichletBC in terms of Te? If so, Zapdos has that (please see `ElectronTemperatureDirichletBC.C` ) All the user has to do is supply the electron temperature in eV.

**Reply by keniley1** [reply]
Haha I honestly had no idea ElectronTemperatureDirichletBC existed! That is indeed what I was talking about. The fact that this exists pretty much dismisses the second point in my post. 

I do still think the idea of a log-mole energy density is confusing to a lot of people (I've had three separate new users email me about it -- what it means when the IC is -20, what the variable actually is, etc), but in the long run I think all of that should be hidden from the user anyway.

**Reply by cticenhour** [reply]
Using "Te" as a material property to preserve derivatives and get some generalization is a good idea. I also wonder if we could get away with starting to strip back some of these log formulations.... Considering AD, automatic scaling, etc., we may have some ways now to simplify code and still get the performance we expect.

**Reply by csdechant** [reply]
I agree that the log-mole form is confusing, but until we have a better way of ensuring that we don't get negative densities or temperature, I think keeping log form is for the better. I also agree that while we are using log form, we should try to keep it "under the hood" as much as we can. Maybe having a DensityIC / ElectronTemperatureIC object would be good for new users?

**Reply by csdechant** [reply]
Sorry, I hit send before seeing @cticenhour comment. I thought the log form was a way to sure positive values of density / energy, not necessarily to have better performance?

**Reply by keniley1** [reply]
Yeah, I would love to get rid of log formulation but it exists to enforce positivity. I think it's actually less performant than a linear formulation both in terms of convergence and computational speed (all those std::exp calls add up!), but you run the risk of getting negative densities in a linear formulation. What we **_really_** need is a finite volume implementation of the drift-diffusion-reaction equations. I think FV is (or at least can be?) positivity preserving, and mathematically it suits the physics we're trying to model here. That's a separate discussion though. And a lot of work :) 


> Maybe having a DensityIC / ElectronTemperatureIC object would be good for new users?

Yeah, I think it would be really beneficial for the user to just be able to add densities, temperature, etc. in physical units. That was part of the goal of the new action I'm working on (although it's been on pause for a while) -- I don't want the user to have to dig through documentation and code just to find the names of all the materials, input parameters, log-mole definitions, etc. they need.  A whole lot of this can be done automatically by the code and make things easier for the user.

**Reply by cticenhour** [reply]
Performance definitely isn't the right term, if I consider it more. I was referring to overall user and developer experience (the answer we get, and the ways we go about getting it, as much as the time it takes to get it). Having non-negative values is part of having the log form, for sure. But scaling was also an aspect of it, since the log-form keeps our variables closer to the same order of magnitude.

AD is helping us with complicated Jacobians, and auto-scaling could remove the need some of manual scaling and re-factoring we perform. But we'd only find out if negative values are still a thing by doing a test without log. Just a thought! 

I like the idea of hiding as many things as we can from the end user for problem setup, to make their experience easier. @csdechant's idea of making custom ICs for setting certain parameters is a great idea in service of that goal.

**Reply by cticenhour** [reply]
@keniley1 Corey is actually working on that as part of an INL internship (using the drift-diffusion-reaction equations as an FV-FE coupling test basis)! So it might come sooner than you think :)

**Reply by lindsayad** [reply]
I don't think perfect Jacobians or automatic scaling will do anything to prevent negative densities. Wiggles are a part of the residuals for a continuous finite element formulation (certainly for an unstabilized one). They are also a part of finite volume formulations as well, depending on the interpolation schemes chosen. However, as @keniley1 is perhaps referencing there are TVD schemes for finite volume (and for DG) that will prevent or at least inhibit formation of new local extrema (wiggles). These TVD schemes are always nonlinear, however, and will also impact nonlinear convergence.

So yea FV/DG is the right discretization scheme for advection dominated problems moving forward. However, it is not a silver bullet. Wiggles will still appear at times, most likely, and convergence may still be difficult ... but overall it should be better.

**Reply by cticenhour** [reply]
You're right @lindsayad, but I was also getting at the fact that we have made many improvements in our formulations (shouldn't have only said "Jacobians" in my previous message), between AD conversion simplifying the code base and @csdechant's MMS work, that we may not be as affected by those wiggles in a catastrophic way as before. But I wouldn't want to venture a guess before trying it in a test.
